### ISSUE TITLE
[no-unnecessary-type-assertion] False positive with null/undefined union type

### ISSUE THREAD

**Issue by codebutler** [typescript-eslint/typescript-eslint#1061]
Looks like the fix for #529 did not cover all edge cases.

```JSON
{
  "rules": {
      "@typescript-eslint/no-unnecessary-type-assertion": "error",
  }
}
```

```TS
const optionalNonNull = (s?: string) => { };
let s: string | null | undefined;

/*
Argument of type 'string | null | undefined' is not assignable to parameter of type 'string | undefined'.
  Type 'null' is not assignable to type 'string | undefined'.
*/
optionalNonNull(s);

/*
This assertion is unnecessary since the receiver accepts the original type of the expression.
*/
optionalNonNull(s!);
```

**Expected Result**

Should not trigger an error.

**Actual Result**

Triggers an error.

**Additional Info**

N/A

**Versions**

```
"@typescript-eslint/eslint-plugin": "^2.3.3",
```

Also tested with master 054df278d6b7064a44b5f78fec453bf9ae6ad281

**Comment by bradzacher** [comment]
#1002 fixes this
